src/python/grpcio_testing/grpc_testing/__init__.py

```diff
@@ -0,0 +1,441 @@
+import collections
+
+import grpc
+import six
+
+
+class ChannelRpc(six.with_metaclass(abc.ABCMeta)):
+    """Fixture for an in-progress RPC invoked by a system under test.
+
+    Enables users to "play a server" for such RPCs by reading requests,
+    supplying responses, and supplying status.
+    """
+
+    @abc.abstractmethod
+    def invocation_metadata(self):
+        """Accesses the metadata passed by the system under test.
+
+        Returns:
+          The invocation metadata supplied by the system under test at
+            RPC-invocation-time.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def initial_metadata(self, initial_metadata):
+        """Accepts the initial metadata to be passed to the system under test.
+
+        Args:
+          initial_metadata: The RPC's initial metadata to be passed to the
+            system under test.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def take_request_as_message(self):
+        """Draws one of the requests added to the RPC by the system under test.
+
+        Successive calls to this method return requests in the same order in
+        which the system under test added them to the RPC.
+
+        Returns:
+          A deserialized request message added to the RPC by the system under
+            test.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def add_responses_as_messages(self, responses):
+        """Accepts responses to be passed to the system under test.
+
+        Args:
+          responses: An iterable of deserialized response messages to be passed
+            to the system under test.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def no_more_requests(self):
+        """Blocks until the system under test has closed the request stream."""
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def cancelled(self):
+        """Blocks until the system under test has cancelled the RPC."""
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def terminate(self, trailing_metadata, code, details):
+        """Accepts the RPC's trailing metadata, status code, and status details.
+
+        This method implicitly closes the RPC's response stream and, from the
+        perspective of the test code that is "playing server", terminates the
+        RPC.
+        """
+        raise NotImplementedError()
+
+
+class ChannelFixture(six.with_metaclass(abc.ABCMeta)):
+    """Fixture for a channel on which a system under test invokes RPCs."""
+
+    @abc.abstractmethod
+    def channel(self):
+        """Accesses the grpc.Channel on which the system under test makes RPCs.
+
+        Returns:
+          The grpc.Channel on which the system under test should make RPCs.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def take_rpc_by_service_and_method_names(self, service_name, method_name):
```

Comment:
What do you think of changing this to ```get_next_rcp()``` and having a service_name and method name field as part of the ChannelRPC?